clarity/src/vm/analysis/type_checker/mod.rs

```diff
@@ -307,34 +375,324 @@ impl FunctionType {
 
-        for (expected_arg, arg) in expected_args.iter().zip(func_args.iter()).into_iter() {
-            match (&expected_arg.signature, arg) {
-                (
-                    TypeSignature::TraitReferenceType(trait_id),
-                    Value::Principal(PrincipalData::Contract(contract)),
-                ) => {
-                    let contract_to_check = db
-                        .load_contract(contract)
-                        .ok_or_else(|| CheckErrors::NoSuchContract(contract.name.to_string()))?;
-                    let trait_definition = db
-                        .get_defined_trait(&trait_id.contract_identifier, &trait_id.name)
-                        .unwrap()
-                        .ok_or(CheckErrors::NoSuchContract(
-                            trait_id.contract_identifier.to_string(),
-                        ))?;
-                    contract_to_check.check_trait_compliance(trait_id, &trait_definition)?;
+        let mut arg_types = Vec::new();
+        for arg in func_args {
+            arg_types.push(self.principal_to_callable_type(arg, 1)?);
+        }
+
+        for (expected_arg, arg_type) in expected_args.iter().zip(arg_types.iter()).into_iter() {
+            inner_type_check_type(
+                db,
+                None,
+                arg_type,
+                &expected_arg.signature,
+                clarity_version,
+                1,
+                &mut LimitedCostTracker::new_free(),
+            )?;
+        }
+        Ok(returns.clone())
+    }
+}
+
+pub fn trait_check_trait_compliance<T: CostTracker>(
+    db: &mut AnalysisDatabase,
+    contract_context: Option<&ContractContext>,
+    actual_trait_identifier: &TraitIdentifier,
+    actual_trait: &BTreeMap<ClarityName, FunctionSignature>,
+    expected_trait_identifier: &TraitIdentifier,
+    expected_trait: &BTreeMap<ClarityName, FunctionSignature>,
+    clarity_version: ClarityVersion,
+    tracker: &mut T,
+) -> CheckResult<()> {
+    // Shortcut for the simple case when the two traits are the same.
+    if actual_trait_identifier == expected_trait_identifier {
+        return Ok(());
+    }
+
+    for (func_name, expected_sig) in expected_trait.iter() {
+        if let Some(func) = actual_trait.get(func_name) {
+            let args_iter = expected_sig.args.iter().zip(func.args.iter());
+            for (expected_type, actual_type) in args_iter {
+                if inner_type_check_type(
+                    db,
+                    contract_context,
+                    actual_type,
+                    expected_type,
+                    clarity_version,
+                    1,
+                    tracker,
+                )
+                .is_err()
+                {
+                    return Err(CheckErrors::IncompatibleTrait(
+                        expected_trait_identifier.clone(),
+                        actual_trait_identifier.clone(),
+                    )
+                    .into());
                 }
```

Comment:
Do you think it makes sense to pull this into another function like `check_functions_compatible(expected, actual)`? In the current implementation, this is `FunctionSignature::check_args_trait_compliance()`. I think it'd make it more obvious the checks that are being applied. I think what we want checked is:

1. Does every argument of `actual` admit the corresponding argument of `expected` (noting the "admission" direction)
2. Does the return type of `expected` admit the return type of `actual`?
3. Is the argument count of `actual` equal to the argument count of `expected`?

Looking at this code segment, I think (3) is missed, and I _think_ the admission direction of (1) is backwards?

Reply:
2. Addressed in f5fd08a19 (including adding tests to catch this).
3. I think this is correct. If it was wrong, `test_trait_with_superset_trait_arg` and `test_trait_with_subset_trait_arg` would fail. Do you see anything wrong with those tests?

Reply:
Refactored in 2345f9d64